### ISSUE TITLE
Demucs Inference for Real Time Streams

### ISSUE THREAD

**Issue by Matheusadler** [facebookresearch/demucs#515]
## ❓ Questions

First of all, congratulations on the excellent work!

Is there already any study or issue addressing the topic of inference for real-time streams?

**Comment by gongouveia** [comment]
@Matheusadler Do you have any further information regarding this topic?

**Comment by CarlGao4** [comment]
No plan to support this
I think it’s not difficult to write a program calling demucs API and separate it chunk by chunk

**Comment by CarlGao4** [comment]
No plan to support this
I think it’s not difficult to write a program calling demucs API and separate it chunk by chunk

**Comment by gongouveia** [comment]
Yes, I can do it. But I would like to have information regarding the demucsv4 based real time latency compared to the demucsv2. 

Will it run in similar latency in the same hardware compared to its older versions?

**Comment by CarlGao4** [comment]
Hmmm...
I think that you want to ask whether your hardware can reach real-time separation.
I've estimated that your graphic card should at least reach 2 TFLOPS on calculating Float32 if you want to separate an audio using a htdemucs (v4) model with 4 submodels (shifts=0, overlap=0.25). Separation speed of hdemucs (v3) models is about 3 times of v4. I haven't tested demucs (v2).
And about latency, demucs doesn't process data like stream actually. You should cut your audio into chunks and pass in the chunks one by one. The data above only ensures that demucs can output the last chunk before the length of last chunk exceeds. So the actual latency is about two times of the length of your chunk. Please also remember to prepare overlap of your chunks

**Comment by gongouveia** [comment]
@CarlGao4 Thank you, that really helps.
Due to the transformer nature I believe that the longer the audio chunk the better is the audio separation.
facebook denoiser used Demucs V2  and it can achieves quasi real time separation.
